src/store/writer.rs

```diff
@@ -85,76 +125,145 @@ impl StoreWriter {
+        let doc_num_bytes = serialized_document.len();
+        VInt(doc_num_bytes as u64).serialize(&mut self.current_block)?;
+        self.current_block.write_all(serialized_document)?;
+        self.num_docs_in_current_block += 1;
+        self.check_flush_block()?;
         Ok(())
     }
 
     /// Stacks a store reader on top of the documents written so far.
     /// This method is an optimization compared to iterating over the documents
     /// in the store and adding them one by one, as the store's data will
     /// not be decompressed and then recompressed.
-    pub fn stack(&mut self, store_reader: &StoreReader) -> io::Result<()> {
-        if !self.current_block.is_empty() {
-            self.write_and_compress_block()?;
-        }
-        assert_eq!(self.first_doc_in_block, self.doc);
-        let doc_shift = self.doc;
-        let start_shift = self.writer.written_bytes() as usize;
+    pub fn stack(&mut self, store_reader: StoreReader) -> io::Result<()> {
+        self.check_flush_block()?;
+        let block = self.get_current_block();
+        self.compressor_sender
+            .send(BlockCompressorMessage::Stack((store_reader, block)))
+            .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
 
-        // just bulk write all of the block of the given reader.
-        self.writer
-            .write_all(store_reader.block_data()?.as_slice())?;
+        Ok(())
+    }
 
-        // concatenate the index of the `store_reader`, after translating
-        // its start doc id and its start file offset.
-        for mut checkpoint in store_reader.block_checkpoints() {
-            checkpoint.doc_range.start += doc_shift;
-            checkpoint.doc_range.end += doc_shift;
-            checkpoint.byte_range.start += start_shift;
-            checkpoint.byte_range.end += start_shift;
-            self.register_checkpoint(checkpoint);
+    /// Finalized the store writer.
+    ///
+    /// Compress the last unfinished block if any,
+    /// and serializes the skip list index on disc.
+    pub fn close(mut self) -> io::Result<()> {
+        let block = self.get_current_block();
+        if !block.is_empty() {
+            self.compressor_sender
+                .send(BlockCompressorMessage::AddBlock(block))
+                .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
         }
+        drop(self.compressor_sender);
+
+        self.compressor_thread_handle
+            .join()
+            .map_err(|err| io::Error::new(io::ErrorKind::Other, format!("{:?}", err)))??;
+
         Ok(())
     }
+}
 
-    fn register_checkpoint(&mut self, checkpoint: Checkpoint) {
-        self.offset_index_writer.insert(checkpoint.clone());
-        self.first_doc_in_block = checkpoint.doc_range.end;
-        self.doc = checkpoint.doc_range.end;
+/// BlockCompressor is separated from StoreWriter, to be run in an own thread
+pub struct BlockCompressor {
+    compressor: Compressor,
+    doc: DocId,
```

Comment:
The previous name `first_doc_in_block` vs `doc` was actually clearer.
Why the rename? 

If it is motivated can you add a comment at the very least?